rts/motoko-rts/src/gc/incremental.rs

```diff
@@ -1,18 +1,424 @@
+            Phase::Pause | Phase::Stop => unreachable!(),
+        }
+    }
+
+    /// Only to be called when the call stack is empty as pointers on stack are not collected as roots.
+    unsafe fn start_marking(&mut self, roots: Roots) {
+        debug_assert!(self.pausing());
+
+        self.state.phase = Phase::Mark;
+        MarkIncrement::start_phase(self.mem, self.state, &mut self.time);
+        let mut increment = MarkIncrement::instance(self.mem, self.state, &mut self.time);
+        increment.mark_roots(roots);
+    }
+
+    unsafe fn mark_completed(&self) -> bool {
+        self.state.phase == Phase::Mark && MarkIncrement::<M>::mark_completed(self.state)
+    }
+
+    unsafe fn check_mark_completion(&mut self, _roots: Roots) {
+        #[cfg(feature = "memory_check")]
+        {
+            sanity_checks::check_memory(
+                self.mem,
+                &mut self.state.partitioned_heap,
+                _roots,
+                sanity_checks::CheckerMode::MarkCompletion,
+            );
+        }
+    }
+
+    unsafe fn start_evacuating(&mut self, roots: Roots) {
+        self.check_mark_completion(roots);
+        debug_assert!(self.mark_completed());
+        MarkIncrement::<M>::complete_phase(self.state);
+        self.state.phase = Phase::Evacuate;
+        EvacuationIncrement::<M>::start_phase(self.state);
+    }
+
+    unsafe fn evacuation_completed(&self) -> bool {
+        self.state.phase == Phase::Evacuate
+            && EvacuationIncrement::<M>::evacuation_completed(self.state)
+    }
+
+    unsafe fn start_updating(&mut self, roots: Roots) {
+        debug_assert!(self.evacuation_completed());
+        EvacuationIncrement::<M>::complete_phase(self.state);
+        self.state.phase = Phase::Update;
+        UpdateIncrement::start_phase(self.state);
+        let mut increment = UpdateIncrement::instance(self.state, &mut self.time);
+        increment.update_roots(roots);
+    }
+
+    unsafe fn updating_completed(&self) -> bool {
+        self.state.phase == Phase::Update && UpdateIncrement::update_completed(self.state)
+    }
+
+    /// Only to be called when the call stack is empty as pointers on stack are not updated.
+    unsafe fn complete_run(&mut self, roots: Roots) {
+        debug_assert!(self.updating_completed());
+        UpdateIncrement::complete_phase(self.state);
+        self.state.phase = Phase::Pause;
+        self.check_update_completion(roots);
+    }
+
+    unsafe fn check_update_completion(&mut self, _roots: Roots) {
+        #[cfg(feature = "memory_check")]
+        {
+            sanity_checks::check_memory(
+                self.mem,
+                &mut self.state.partitioned_heap,
+                _roots,
+                sanity_checks::CheckerMode::UpdateCompletion,
+            );
+        }
+    }
+}
+
+/// Write barrier to be called BEFORE a potential overwrite of a pointer value.
+/// `overwritten_value` (skewed if a pointer) denotes the value that will be overwritten.
```

Comment:
This is new value, right? Not the current contents of location, before the write.

Reply:
No, this should be the old value (maybe I should rename that it is clearer). This is important because of the snapshot-at-the-beginning principle: (At least) all objects are marked that have been reachable at the beginning of the GC run. If a value is overwritten, we could miss that original reference and need to mark it to be sure we have a super-set of the initial snapshot marked.

Reply:
Ok, I guess that's the snap-shot at the beginning thing. Makes sense.